src/main.cpp

```diff
@@ -4073,6 +4076,69 @@ bool CVerifyDB::VerifyDB(const CChainParams& chainparams, CCoinsView *coinsview,
     return true;
 }
 
+bool RewindBlockIndex(const Consensus::Params& params)
+{
+    LOCK(cs_main);
+
+    int nHeight = 1;
+    while (nHeight <= chainActive.Height()) {
+        if (IsWitnessEnabled(chainActive[nHeight - 1], params) && !(chainActive[nHeight]->nStatus & BLOCK_OPT_WITNESS)) {
+            break;
+        }
+        nHeight++;
+    }
+
+    // nHeight is now the height of the first insufficiently-validated block, or tipheight + 1
+    CValidationState state;
+    CBlockIndex* pindex = chainActive.Tip();
+    while (chainActive.Height() >= nHeight) {
+        if (!DisconnectTip(state, params, true)) {
+            return error("RewindBlockIndex: unable to disconnect block at height %i", pindex->nHeight);
+        }
+        // Occasionally flush state to disk.
+        if (!FlushStateToDisk(state, FLUSH_STATE_PERIODIC))
+            return false;
+    }
+
+    // Reduce validity flag and have-data flags.
+    // We do this after actual disconnecting, otherwise we'll end up writing the lack of data
+    // to disk before writing the chainstate, resulting in a failure to continue if interrupted.
+    for (BlockMap::iterator it = mapBlockIndex.begin(); it != mapBlockIndex.end(); it++) {
+        CBlockIndex* pindexIter = it->second;
+        if (IsWitnessEnabled(pindexIter->pprev, params) && !(pindexIter->nStatus & BLOCK_OPT_WITNESS)) {
+            // Reduce validity
+            pindexIter->nStatus = std::min<unsigned int>(pindexIter->nStatus & BLOCK_VALID_MASK, BLOCK_VALID_TREE) | (pindexIter->nStatus & ~BLOCK_VALID_MASK);
+            // Remove have-data flags.
+            pindexIter->nStatus &= ~(BLOCK_HAVE_DATA | BLOCK_HAVE_UNDO);
+            // Remove storage location.
+            pindexIter->nFile = 0;
+            pindexIter->nDataPos = 0;
+            pindexIter->nUndoPos = 0;
+            // Remove various other things
+            pindexIter->nTx = 0;
+            pindexIter->nChainTx = 0;
+            pindexIter->nSequenceId = 0;
+            // Make sure it gets written.
+            setDirtyBlockIndex.insert(pindexIter);
+            // Update setBlockIndexCandidates
+            setBlockIndexCandidates.erase(pindexIter);
+            if ((!IsWitnessEnabled(pindexIter->pprev->pprev, params) || (pindexIter->pprev->nStatus & BLOCK_OPT_WITNESS)) && pindexIter->pprev->IsValid(BLOCK_VALID_TRANSACTIONS) && pindexIter->pprev->nChainTx) {
```

Comment:
Looking at `pprev` here is confusing, since you need to repeat the condition in the `if()` above to make this work.  I think this is right, but perhaps it'd be easier to read if, in this loop, we just consider whether to add the pindexIter we're looking at to setBlockIndexCandidates, say by rewriting this as follows:
```
    ....
    setBlockIndexCandidates.erase(pindexIter);
} else if (pindexIter->IsValid(BLOCK_VALID_TRANSACTIONS) && pindexIter->nChainTx) {
    setBlockIndexCandidates.insert(pindexIter);
}
```
This is inefficient from the perspective of adding more entries than is strictly needed to `setBlockIndexCandidates`, and therefore the prune below will take a little longer, but that seems like a small effect?